staging/vhost-device-gpu/src/vhu_gpu.rs

```diff
@@ -551,30 +470,37 @@ impl VhostUserGpuBackend {
         virtio_gpu: &mut VirtioGpu,
         vring: &VringRwLock,
     ) -> Result<()> {
-        // Collect all pending requests
-        debug!("Processing control queue");
-        let Some(ref atomic_mem) = self.mem else {
-            return Err(Error::NoMemoryConfigured.into());
-        };
-        let requests: Vec<_> = vring
+        let mem = self.mem.as_mut().unwrap().memory().into_inner();
+        let desc_chains: Vec<_> = vring
             .get_mut()
             .get_queue_mut()
-            .iter(atomic_mem.memory())
-            .map_err(|_| Error::DescriptorNotFound)?
+            .iter(mem.clone())
+            .map_err(Error::CreateIteratorDescChain)?
             .collect();
 
-        let mem: GuestMemoryMmap = (*atomic_mem.memory().into_inner()).clone();
+        let mut signal_used_queue = false;
+        for desc_chain in desc_chains {
+            let head_index = desc_chain.head_index();
+            let mut reader = desc_chain
+                .clone()
+                .reader(&mem)
+                .map_err(Error::CreateReader)?;
+            let mut writer = desc_chain.writer(&mem).map_err(Error::CreateWriter)?;
+
+            self.process_control_queue_chain(
+                virtio_gpu,
+                vring,
+                head_index,
+                &mut reader,
+                &mut writer,
+                &mut signal_used_queue,
+            )?;
+        }
 
-        if self.process_requests(requests, virtio_gpu, vring).is_ok() {
-            // Send notification once all the requests are processed
-            debug!("Sending processed request notification");
-            vring
-                .signal_used_queue()
-                .map_err(|_| Error::NotificationFailed)?;
-            debug!("Notification sent");
+        if signal_used_queue {
+            vring.signal_used_queue().map_err(Error::SignalUsedQueue)?;
```

Comment:
here, why not adapt it to how the check was done like in process_requests i.e  ```if self.process_control_queue_chain().is_ok() { ....``` rather than using this bool `signal_used_queue`

Reply:
The `if self.process_control_queue_chain().is_ok()` swallowed errors (not reported them), so that is obviously not good, but that could be solved by just using `match` instead.

My reason for using an explicit variable `signal_used_queue` is that, if we don't call `vring.add_used(...)` we don't need to call ` vring.signal_used_queue()`.  But `.is_ok()` could be true, even if we never called the  `vring.add_used(...)`.  (From source code, it looks like both libkrun and crossvm, don't do `signal_used_queue` when they don't do `add_used`)

Another option could be to return a `Result<bool>` from  `process_control_queue_chain` that could be maybe a bit cleaner. But also depends how we want to eventual handle errors - If we want to stop at first error, or just `log::error!` and try to continue processing other descriptor chains.

Reply:
that's weird, I didn't know `self.process_control_queue_chain()` wasn't reporting the errors. Good catch btw. Regarding how we want to handle the error, i think it's better to stop at the first error encountered, because if one descriptor is not process it could cause problems for the others what do you think?

Reply:
Yes, makes sense.  Actually this even happened to us before this PR, we were receiving `TransferToHost2d`, but never properly parsed the `ResourceAttachBacking`. And `TransferToHost2d` doesn't make sense without `ResourceAttachBacking`.